### PR TITLE
Customized `tracing` for `tide::Middleware` logging

### LINKED ISSUE #453: 🪵 The 🪵 Great 🪵 Log 🪵 Issue 🪵
Logs should be:

1. Fast.
2. Timestamped.
3. JSON.
4. Correlatable.

[json_env_logger](https://lib.rs/crates/json_env_logger) seems sensible from an import-a-crate-make-go perspective, env_logger is already being used in a bunch of places and replacing it with json\_env\_logger is largely a find-and-replace matter - then the long tail of `%s/eprintln/error/g`.

It doesn't do everything, but... it might be able to once the formatting options are considered.

I have a much longer rant about logging [here](https://yaleman.org/post/2021/2021-05-22-the-great-log-post-/) if you want to read it. It got pretty long, so ... let's start with ^ that.

### PR DESCRIPTION
**Pull request by QnnOkabayashi**
Fixes #453 

- [x] cargo fmt has been run
- [ ] cargo clippy has been run
- [ ] cargo test has been run and passes
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

Started the process of integrating my custom `tracing` subscriber into `kanidmd`, specifically with `tide` requests and collecting logs from request all the way to response.

I need help figuring out how to run it in a way that would produce logs. I tried running this command from awhile back, but it doesn't actually produce logs so I don't know if I'm initializing the subscriber in the right place.
```
cargo run -- recover_account -c ./server.toml -n admin
```

### CODE REVIEW

File: kanidmd/src/lib/core/https/mod.rs [hunk 1]
```diff
@@ -1173,6 +1177,13 @@ pub fn create_https_server(
     qe_w_ref: &'static QueryServerWriteV1,
     qe_r_ref: &'static QueryServerReadV1,
 ) -> Result<(), ()> {
+    let (log_tx, log_rx) = unbounded::<TreeProcessor<KanidmEventTag>>();
+    tracing::subscriber::set_global_default(TreeSubscriber::pretty(log_tx))
+        .expect("Global subscriber already set, this is a bug");
```
Comment (Firstyear) [inline kanidmd/src/lib/core/https/mod.rs:1182]: Needs more '🚨 ' emoji.

File: kanidmd/src/lib/core/https/mod.rs [hunk 2]
```diff
@@ -1173,6 +1177,13 @@ pub fn create_https_server(
...
+    let middleware = TreeMiddleware::with_stdout();
```
Comment (Firstyear) [inline kanidmd/src/lib/core/https/mod.rs:1183]: Probably needs a better name than just middleware.

File: kanidmd/src/lib/tracing_tree/subscriber.rs [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+                    logs,
+                })
+                .expect("Processing channel has been closed, cannot log events."),
+        }
+    }
+
+    fn alarm(event: &TreeEvent<E>, maybe_scope: Option<ScopeFromRoot<Registry>>) -> fmt::Result {
+        // This is an emergency and should be sent to the admin immediately
+        // Hence why we are formatting in the working thread
+        let mut writer = event.timestamp.to_rfc3339();
```
Comment (Firstyear) [inline kanidmd/src/lib/tracing_tree/subscriber.rs:175]: What type is writer here? I'm curious how this works :)

**Review by Firstyear — APPROVED** [review]
I've seen a lot of your early work on this, so I think it looks really good already.

**Comment by Firstyear** [comment]
Added @yaleman to have a look here. @QnnOkabayashi do you mind posting an example of both a json and pretty text log for him to see?

File: kanidmd/src/lib/core/https/mod.rs (hunk 2 above)
Comment (yaleman) [reply kanidmd/src/lib/core/https/mod.rs:1183]: ^ This, given that other middlewares might happen in the future.

File: kanidmd/src/lib/tracing_tree/event_tag.rs [hunk 4]
```diff
@@ -0,0 +1,104 @@
+use super::subscriber::EventTagSet;
+use std::convert::TryFrom;
+
+#[derive(Debug, Clone, Copy)]
+pub enum KanidmEventTag {
```
Comment (yaleman) [inline kanidmd/src/lib/tracing_tree/event_tag.rs:5]: Is there reason you're reimplementing LogTag in `src/lib/audit.rs` which seems to have very similar values?

File: kanidmd/src/lib/tracing_tree/formatter.rs [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+use tracing::Level;
+use tracing_serde::AsSerde;
+
+#[derive(Clone, Copy, Debug)]
+pub enum LogFmt {
+    Json,
+    Pretty,
+}
+
+const EVENT_UUID: &str = "00000000-0000-0000-0000-000000000000";
```
Comment (yaleman) [inline kanidmd/src/lib/tracing_tree/formatter.rs:14]: Const's should probably be in that place where @Firstyear documents past decisions about numbers, and I'm not sure the default should be all-0's - isn't that the Anonymous user uuid?

**Review by yaleman — CHANGES_REQUESTED** [review]
Looks pretty good so far, per the screenshots I saw via email - other than the comments inline above 😄 thanks! I'd love to see current screenshots of the various things with the current code.

File: kanidmd/src/lib/tracing_tree/formatter.rs (hunk 5 above)
Comment (Firstyear) [reply kanidmd/src/lib/tracing_tree/formatter.rs:14]: I'm not sure here, the all 0000 is pretty clear, but it is also likely to cause confusion if someone is grepping for anonymous. https://github.com/kanidm/kanidm/tree/master/kanidmd/src/lib/constants

File: kanidmd/src/lib/tracing_tree/event_tag.rs (hunk 4 above)
Comment (Firstyear) [reply kanidmd/src/lib/tracing_tree/event_tag.rs:5]: i think he plans to drop the content of audit.rs, which I'm okay with.

File: kanidmd/src/lib/tracing_tree/subscriber.rs (hunk 3 above)
Comment (QnnOkabayashi) [reply kanidmd/src/lib/tracing_tree/subscriber.rs:175]: It's just a `String`, I did it like that since `to_rfc3339` returns a `String` and I would've appended it to an empty `String` anyways. I'm going to make this clearer in the next commit, though.

File: kanidmd/src/lib/tracing_tree/event_tag.rs (hunk 4 above)
Comment (QnnOkabayashi) [reply kanidmd/src/lib/tracing_tree/event_tag.rs:5]: I did it so I could have similar tags on my test branch, but either this or the `LogTag` type is going to go. What's important is we have some `enum` type that implements `tracing_tree::subscriber::EventTagSet`

**Comment by QnnOkabayashi** [comment]
> Added @yaleman to have a look here. @QnnOkabayashi do you mind posting an example of both a json and pretty text log for him to see?

Here ya go @yaleman :)

### Pretty
```log
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.248419+00:00 TRACE    try_from_entry_ro [ 620µs | 8.817% / 100.000% ]
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.248485+00:00 TRACE    ┝━ server::internal_search [ 565µs | 12.198% / 91.183% ]
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.248504+00:00 INFO     │  ┝━ 💬 [filter.info]: Some filter info...
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.248553+00:00 TRACE    │  ┕━ server::search [ 490µs | 6.062% / 78.985% ]
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.248567+00:00 TRACE    │     ┝━ be::search [ 442µs | 6.741% / 71.310% ]
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.248578+00:00 TRACE    │     │  ┝━ be::search -> filter2idl [ 385µs | 16.593% / 62.097% ]
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.248594+00:00 TRACE    │     │  │  ┝━ be::idl_arc_sqlite::get_idl [ 25.3µs | 4.086% ]
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.248603+00:00 INFO     │     │  │  │  ┕━ 💬 [filter.info]: Some filter info...
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.248702+00:00 TRACE    │     │  │  ┕━ be::idl_arc_sqlite::get_idl [ 257µs | 41.418% ]
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.248710+00:00 ERROR    │     │  │     ┝━ 🚨 [admin.error]: Oopsies, an admin error occurred :)
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.248717+00:00 DEBUG    │     │  │     ┝━ 🐛 [debug]: An untagged debug log
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.248721+00:00 ERROR    │     │  │     ┕━ 🚨 [error]: there's been a big mistake | alive: false | status: "very sad"
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.248985+00:00 TRACE    │     │  ┕━ be::idl_arc_sqlite::get_identry [ 15.3µs | 2.473% ]
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.248992+00:00 INFO     │     │     ┝━ 🔐 [security.critical]: A security critical log
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.248999+00:00 INFO     │     │     ┕━ 🔓 [security.access]: A security access log
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.249025+00:00 TRACE    │     ┕━ server::search<filter_resolve> [ 10.0µs | 1.613% ]
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.249033+00:00 WARN     │        ┕━ 🚧 [filter.warn]: Some filter warning lol
257dc00d-2507-4e05-a639-9441f00e3cf0 2021-07-30T19:13:40.249077+00:00 TRACE    ┕━ 📍 [trace]: We finished!
```

### Json
```json
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923358+00:00","level":"TRACE","message":"try_from_entry_ro","log-type":"span","nanos-nested":579875,"nanos-total":617916}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923427+00:00","level":"TRACE","message":"server::internal_search","log-type":"span","nanos-nested":500542,"nanos-total":579875}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923447+00:00","level":"INFO","message":"Some filter info...","log-type":"event","tag":"filter.info","spans":["try_from_entry_ro","server::internal_search"]}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923497+00:00","level":"TRACE","message":"server::search","log-type":"span","nanos-nested":456958,"nanos-total":500542}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923513+00:00","level":"TRACE","message":"be::search","log-type":"span","nanos-nested":397542,"nanos-total":445375}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923526+00:00","level":"TRACE","message":"be::search -> filter2idl","log-type":"span","nanos-nested":287542,"nanos-total":380709}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923544+00:00","level":"TRACE","message":"be::idl_arc_sqlite::get_idl","log-type":"span","nanos-nested":0,"nanos-total":25083}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923554+00:00","level":"INFO","message":"Some filter info...","log-type":"event","tag":"filter.info","spans":["try_from_entry_ro","server::internal_search","server::search","be::search","be::search -> filter2idl","be::idl_arc_sqlite::get_idl"]}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923638+00:00","level":"TRACE","message":"be::idl_arc_sqlite::get_idl","log-type":"span","nanos-nested":0,"nanos-total":262459}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923648+00:00","level":"ERROR","message":"Oopsies, an admin error occurred :)","log-type":"event","tag":"admin.error","spans":["try_from_entry_ro","server::internal_search","server::search","be::search","be::search -> filter2idl","be::idl_arc_sqlite::get_idl"]}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923656+00:00","level":"DEBUG","message":"An untagged debug log","log-type":"event","tag":null,"spans":["try_from_entry_ro","server::internal_search","server::search","be::search","be::search -> filter2idl","be::idl_arc_sqlite::get_idl"]}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923661+00:00","level":"ERROR","message":"there's been a big mistake","log-type":"event","tag":null,"spans":["try_from_entry_ro","server::internal_search","server::search","be::search","be::search -> filter2idl","be::idl_arc_sqlite::get_idl"]}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923931+00:00","level":"TRACE","message":"be::idl_arc_sqlite::get_identry","log-type":"span","nanos-nested":0,"nanos-total":16833}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923939+00:00","level":"INFO","message":"A security critical log","log-type":"event","tag":"security.critical","spans":["try_from_entry_ro","server::internal_search","server::search","be::search","be::idl_arc_sqlite::get_identry"]}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923946+00:00","level":"INFO","message":"A security access log","log-type":"event","tag":"security.access","spans":["try_from_entry_ro","server::internal_search","server::search","be::search","be::idl_arc_sqlite::get_identry"]}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923977+00:00","level":"TRACE","message":"server::search<filter_resolve>","log-type":"span","nanos-nested":0,"nanos-total":11583}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.923986+00:00","level":"WARN","message":"Some filter warning lol","log-type":"event","tag":"filter.warn","spans":["try_from_entry_ro","server::internal_search","server::search","server::search<filter_resolve>"]}
{"uuid":"6772000d-fae1-41b2-a96e-17cbf5e71d13","timestamp":"2021-07-30T19:16:09.924025+00:00","level":"TRACE","message":"We finished!","log-type":"event","tag":null,"spans":["try_from_entry_ro"]}
```

Does this help?

**Comment by QnnOkabayashi** [comment]
These commits reflect the start of the integration process. I currently just added logs for logging in, and they look great so far! Let me know if you guys have any feedback :)

File: kanidmd/src/lib/actors/v1_read.rs [hunk 6]
```diff
@@ -154,6 +162,7 @@ impl QueryServerReadV1 {
             .await
             .and_then(|r| idm_auth.commit(&mut audit).map(|_| r));
 
+        security_info!("Sending auth result -> {{why is this thing so large}}" /*, res*/);
```
Comment (yaleman) [inline kanidmd/src/lib/actors/v1_read.rs:165]: I assume this is a testing thing? (also all the double-ups of `admin_error` and `ladmin_error`)

Comment (QnnOkabayashi) [reply kanidmd/src/lib/actors/v1_read.rs:165]: > I assume this is a testing thing? (also all the double-ups of `admin_error` and `ladmin_error`)

Yeah, I omitted `res` for testing purposes because it was polluting the look of the log which is what I'm trying to get a sense for right now.

The double ups are because I'm just adding my logs in parallel right now, the `ladmin_error`'s will be removed at a later time :)

File: kanidmd/src/lib/idm/server.rs [hunk 7]
```diff
@@ -149,6 +150,7 @@ impl IdmServer {
         origin: String,
         // ct: Duration,
     ) -> Result<(IdmServer, IdmServerDelayed), OperationError> {
+        let _entered = trace_span!("IdmServer::new").entered();
```
[the remark this answers is not in the document]
Comment (QnnOkabayashi) [reply kanidmd/src/lib/idm/server.rs:153]: `trace_span!()` doesn't require specifying the target, it only requires a name for the span,. I just happened to arbitrarily chose to be the the name of the function because I thought it was the most descriptive.

[the remark this answers is not in the document]
Comment (yaleman) [reply kanidmd/src/lib/idm/server.rs:153]: So, what's the functional difference between `trace_span` and `span`, when we're using it to trace the function the same way span does? 😄

[the remark this answers is not in the document]
Comment (QnnOkabayashi) [reply kanidmd/src/lib/idm/server.rs:153]: With `span`, you have the specify the level, and `trace_span` is just shorthand for specifying `span` at the trace level.

https://docs.rs/tracing/0.1.26/tracing/index.html#shorthand-macros

[the remark this answers is not in the document]
Comment (yaleman) [reply kanidmd/src/lib/idm/server.rs:153]: Thanks 😄

**Comment by Firstyear** [comment]
@QnnOkabayashi I'm going to have a play with this later, but is there anything stopping us from merging it? Large changes like this are hard to coordinate sometimes, with rebasing etc. So if we want we can merge it in smaller chunks as we go, especially since we are 2 months away from a release. What do you think? 

IE merge in this with the middleware, then next step is do some testing of it with some parts, then we start to replace out auditscope in batches. Makes it a bit more manageble for us as humans :)

**Comment by Firstyear** [comment]
tl;dr I'm okay to commit something that's a draft and has mistakes because that way we can all learn from it and improve :)

**Review by yaleman — APPROVED** [review]

**Comment by QnnOkabayashi** [comment]
Yeah, I have this in a decently merge-able state now. I have the logs going to a file for the most part, and have full parallel integration in `idl_sqlite` and `idl_arc_sqlite` so far. Let me know how it looks!

**Comment by Firstyear** [comment]
Can you rebase it to latest master, and make it log to stdout by default? Otherwise ready to merge :)

**Comment by QnnOkabayashi** [comment]
> Can you rebase it to latest master, and make it log to stdout by default? Otherwise ready to merge :)

Finished a second round of integration, going to work on rebasing and preparing to merge first thing Monday morning 👍

**Comment by QnnOkabayashi** [comment]
This isn't ready for merging yet, as there are a bunch of unit tests I'm still struggling to make pass.

The issue is that global subscriber lasts between tests, so I can't just reassign it between tests (which I feel like I should be able to). My idea to resolve this was to just ignore errors from trying to reassign it, but this means that once the first test finishes, it will drop its processing thread, the only receiver of the global subscriber, therefore _disabling_ the global subscriber from sending any more logs without panicking.

Ideas I've considered:
1. Find a way for the current thread to borrow and use the receiver of the global subscriber's sender. Unfortunately, this would require an `Arc<Mutex<UnboundReceiver<_>>>` which I'd really like to avoid.
2. Give the global subscriber a new sender for you have a fresh thread. This doesn't work because a) multiple threads? b) I only have immutable access to the global subscriber once set by using `downcast_ref`.
3. Have a global formatter/processor to avoid the sender/receiver ever being dropped. I don't think you can have a `'static`  thread though, especially on a tokio runtime.
4. Manually flush the threads like `AuditScope`. This also doesn't work because I can't mutably access the global subscriber.

I'll likely go for the first option because the mutex will probably be used like once in practice, but will make the unit tests pass. Very annoying :(

My main concern is whether a solution is necessary for this problem outside of a testing environment. Won't we only have to set a global subscriber once in practice?

**Comment by Firstyear** [comment]
I feel like this must be a problem that's been encountered before you :) 

I think the obvious solution here is rather than setting a default global subscriber in tests, you set the "log" feature on tracing, and then env_logger will capture and display information instead. 

Then in tests that actually use the server core proper, the subscriber can be set through macros.rs for those setups.

Does that help? Or am I missing something?

**Comment by QnnOkabayashi** [comment]
> I feel like this must be a problem that's been encountered before you :)
> 
> I think the obvious solution here is rather than setting a default global subscriber in tests, you set the "log" feature on tracing, and then env_logger will capture and display information instead.

Would this mean that no `tracing` logs would show up in tests? Because my subscriber currently relies on having a processing thread.
> 
> Then in tests that actually use the server core proper, the subscriber can be set through macros.rs for those setups.

I like this idea a lot!
> 
> Does that help? Or am I missing something?

Another idea I had was to abstract my subscriber over a `Processor` trait implemented by a `TestProcessor`, which just processes logs in the same thread and an `ExecProcessor`, which holds an `UnboundedSender<_>` which will be used to send unprocessed logs to another thread for release builds. I think this would work because even though processing and formatting logs on the main thread is bad for performance, which is what we're looking at during tests, it would only ever happen at the highest level between root spans, which isn't in the middle of any critical performance areas.

**Comment by QnnOkabayashi** [comment]
Ended up implementing the above idea, where the `TreeSubscriber` takes a `Processor` as a generic type, an now all the test cases pass :). I also patched a bug where event fields in JSON formatting were ignored.

Ready for merging!

File: kanidmd/src/lib/be/idl_sqlite.rs [hunk 8]
```diff
@@ -2,6 +2,7 @@ use crate::audit::AuditScope;
 use crate::be::{BackendConfig, IdList, IdRawEntry, IdxKey, IdxSlope};
 use crate::entry::{Entry, EntryCommitted, EntrySealed};
 use crate::value::{IndexType, Value};
+use crate::{admin_error, admin_info, admin_warn, filter_error, spanned};
```
Comment (Firstyear) [inline kanidmd/src/lib/be/idl_sqlite.rs:5]: You should consider adding these to prelude instead so you don't need to import them everywhere. :)

File: kanidmd/src/lib/entry.rs [hunk 9]
```diff
@@ -1293,6 +1293,9 @@ impl Entry<EntrySealed, EntryCommitted> {
         }
     }
 
+    // ! TRACING INTEGRATED
+    // Why is this returning a `Result<Self, ()>` when we could just do `Option<Self>`
```
Comment (Firstyear) [inline kanidmd/src/lib/entry.rs:1297]: It doesn't XD

I'm going to fix it this week, I have 4 clippy lints left that need fixing and this is one.

**Review by Firstyear — APPROVED** [review]
Only a few minor comments, I think we'll be pretty close to swapping completely to tracing after this!

**Comment by Firstyear** [comment]
Absolutely *AMAZING* work @QnnOkabayashi - I know there are still plenty of rough edges and more to do, but this is a huge step forward and a massive accomplishment. WELL DONE! 🎉
